File: typescript/Organization.ts
import Position from './Position';
import Name from './Name';
import Employee from './Employee';

abstract class Organization {
  private root: Position;
  private identifier:number=1;
  constructor() {
    this.root = this.createOrganization();
  }

  protected abstract createOrganization(): Position;

  printOrganization = (position: Position, prefix: string): string => {
    let str = `${prefix}+-${position}\n`;
    for (const p of position.getDirectReports()) {
      str = str.concat(this.printOrganization(p, `${prefix}  `));
    }
    return str;
  };

  // Hire the given person as an employee in the position that has that title
  // Return the newly filled position or undefined if no position has that title
  hire(person: Name, title: string): Position | undefined {
    // your code here
   return this.hireRecursion(person,title,this.root);
  }

  private hireRecursion(person: Name, title: string,position:Position):Position | undefined 
  {
    if(position.getTitle()===title)
    {
      position.setEmployee(new Employee(this.identifier++,person));
      return position;
    }
    position.getDirectReports().forEach(element => {
      var result= this.hireRecursion(person,title,element);
      if(result!=undefined)
       return result;
    });
    return undefined;
  }

  toString = () => this.printOrganization(this.root, '');
}

export default Organization;

File: dotnet/MyOrganization/Organization.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MyOrganization
{
    internal abstract class Organization
    {
        private Position root;
        private static int _id = 0;
        public Organization()
        {
            root = CreateOrganization();
        }

        protected abstract Position CreateOrganization();

        /**
         * hire the given person as an employee in the position that has that title
         * 
         * @param person
         * @param title
         * @return the newly filled position or empty if no position has that title
         */
        public Position? Hire(Name person, string title)
        {
            return Hire(person, title, root);
        }
        private Position? Hire(Name person, string title, Position position)
        {
            if (position == null)
                return null;
            if (position.GetTitle() == title)
            {
                _id++;
                position.SetEmployee(new Employee(_id, person));
                return position;
            }
            else
            {
                foreach (var item in position.GetDirectReports())
                {
                    var pos = Hire(person, title, item);
                    if (pos != null)
                        return pos;
                }
            }
            return null;
        }

        override public string ToString()
        {
            return PrintOrganization(root, "");
        }

        private string PrintOrganization(Position pos, string prefix)
        {
            StringBuilder sb = new StringBuilder(prefix + "+-" + pos.ToString() + "\n");
            foreach (Position p in pos.GetDirectReports())
            {
                sb.Append(PrintOrganization(p, prefix + "\t"));
            }
            return sb.ToString();
        }
    }
}
